connection bus integration tests

fhill2 opened the pull request:
I noticed there is limited test coverage for the MessageBus, Connection and Dispatcher thread, and wanted to share an implementation that can improve the coverage.

Here is an implementation (WIP) similar to what i'm using to perform integration tests against my own client.

TWS / Gateway is mocked and run on a thread, The IB reconnect routine is simulated and supports returning responses when requests are received.  The behaviour of the server can be configured at the start of the test by declaring the request & responses vectors similar how the `MessageBusStub` can be configured for each test.

The  difference between this implementation and the current `MessageBusStub`  is the ability to schedule the handshake, and closing the socket.

```rust
   let messages = VecDeque::from(vec![
        Message::handshake("API\0\0\0\0\tv100..173", vec!["173\020250323 22:21:01 Greenwich Mean Time\0"]),
        Message::packet("71\02\028\0\0", vec!["15\01\0DU1234567\0", "9\01\01\0"]),
        // do not send a response back to keep the client waiting...
        Message::packet("9\08\09000\00\0AAPL\0STK\0\00\0\0\0SMART\0\0USD\0\0\00\0\0\0", vec![]),
        Message::Eof,
        Message::handshake("API\0\0\0\0\tv100..173", vec!["173\020250323 22:21:01 Greenwich Mean Time\0"]),
        Message::packet("71\02\028\0\0", vec!["15\01\0DU1234567\0", "9\01\01\0"]),
        Message::packet(
            "9\08\09001\00\0AAPL\0STK\0\00\0\0\0SMART\0\0USD\0\0\00\0\0\0",
            vec![&format!("10\09001\0{}", AAPL_CONTRACT), "52\01\09001\0"],
        ),
    ]);
```

By structuring the tests this way the behaviour of the client before, during, after a disconnect can be tested. As the server or stream is mocked instead of the MessageBus, all modules are tested.

Passing  a mocked TcpStream to the Connection then performing tests can be another solution.

Any thoughts on the best approach to test these modules, and if this implementation can work for the library?

wboayue reviewed (commented):
Good eye. This was a gap I had intended to close at some point.

Mocking the server is also, in my opinion, the best approach to increase coverage for the `MessageBus`. There are libraries like `mockito` to help generate mock servers, but it's geared towards a specific use cases. I also think the custom mock you’re suggesting will result in code that’s easier to follow.

Overall, I think this is the right approach. We just need to refine the code a bit. I like the use of enumerated `Message` types for the mock messages. I prefer using structs for things like `MockServer` so there’s a clean place to attach configuration and cleanup logic. We'll also need to ensure that the tests can still run in parallel which  looks to be handled by the use of dynamically allocated ports.

fhill2 commented:
That's great, I'll refine this and implement the following over the next week:
- use structs for `MockServer`
- add some additional tests using the mock server.
- remove the temporary sleep() from the tests. 
- teardown the mock server correctly on test end.

fhill2 commented on src/messages.rs:
```diff
@@ -349,6 +352,50 @@ impl ToField for OutgoingMessages {
+    min_version: i32,
+    max_version: i32,
+}
+impl RequestHandshake {
```
I've created a different request type for the RequestHandshake so we can create handshake bytestrings to test with more easily

fhill2 commented on src/messages.rs:
```diff
@@ -349,6 +352,50 @@ impl ToField for OutgoingMessages {
+        format!("v{}..{}", self.min_version, self.max_version)
+    }
+
+    pub fn packet(&self) -> String {
```
I've moved the length encoding from the Connection to the Message as there is `encode()` there already, and the method to produce the final bytes relies on the type of Message.

fhill2 commented:
@wboayue how should we handle waiting in the tests for the client to reconnect? Currently I can't find a reliable way to do this.

I've tried waiting based on if a client request method (e.g contract_details()) returns `Err::ConnectionReset` - however I don't think this error is being returned to the test correctly.

Based on the [MessageBus reset() method](https://github.com/wboayue/rust-ibapi/blob/8f2ab14beee94172e81404ae8cb040afbbac2923/src/transport.rs#L197) - the bus should notify the waiters with an Error::ConnectionReset after reconnection.

Below are logs from a test using the mock server, the actions are:
- connect
- contract_details() Send Receive
- disconnect
- contract_details() Send
- reconnect Success <-- should notify all request waiters with Error::ConnectionReset 
- contract_details() Receive <-- empty vec is returned here instead (atleast this is the case for contract_details()).  it should return Error::ConnectionReset


<details>
  <summary>Mock Server Test Logs</summary>

```txt
  [2025-04-01T21:53:42Z DEBUG ibapi::transport] -> "API\0\0\0\0\tv100..173"
[2025-04-01T21:53:42Z DEBUG ibapi::transport::connection::tests] mock <- "v100..173"
[2025-04-01T21:53:42Z DEBUG ibapi::transport::connection::tests] mock -> "\0\0\0+173\020250323 22:21:01 Greenwich Mean Time\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport] <- "173\020250323 22:21:01 Greenwich Mean Time\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport] -> "\0\0\0\t71\02\028\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport::connection::tests] mock <- "71\02\028\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport::connection::tests] mock -> "\0\0\0\u{10}15\01\0DU1234567\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport::connection::tests] mock -> "\0\0\0\u{7}9\01\01\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport] <- "15\01\0DU1234567\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport] <- "9\01\01\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::client] send_message(9000, RequestMessage { fields: ["9", "8", "9000", "0", "AAPL", "STK", "", "0", "", "", "SMART", "", "USD", "", "", "0", "", ""] })
[2025-04-01T21:53:42Z DEBUG ibapi::transport] -> "\0\0\0*9\08\09000\00\0AAPL\0STK\0\00\0\0\0SMART\0\0USD\0\0\00\0\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport::connection::tests] mock <- "9\08\09000\00\0AAPL\0STK\0\00\0\0\0SMART\0\0USD\0\0\00\0\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport::connection::tests] mock -> "\0\0\u{4}r10\09000\0AAPL\0STK\0\00\0\0SMART\0USD\0AAPL\0NMS\0NMS\0265598\00.01\0\0ACTIVETIM,AD,ADDONT,ADJUST,ALERT,ALGO,ALLOC,AON,AVGCOST,BASKET,BENCHPX,CASHQTY,COND,CONDORDER,DARKONLY,DARKPOLL,DAY,DEACT,DEACTDIS,DEACTEOD,DIS,DUR,GAT,GTC,GTD,GTT,HID,IBKRATS,ICE,IMB,IOC,LIT,LMT,LOC,MIDPX,MIT,MKT,MOC,MTL,NGCOMB,NODARK,NONALGO,OCA,OPG,OPGREROUT,PEGBENCH,PEGMID,POSTATS,POSTONLY,PREOPGRTH,PRICECHK,REL,REL2MID,RELPCTOFS,RPI,RTH,SCALE,SCALEODD,SCALERST,SIZECHK,SMARTSTG,SNAPMID,SNAPMKT,SNAPREL,STP,STPLMT,SWEEP,TRAIL,TRAILLIT,TRAILLMT,TRAILMIT,WHATIF\0SMART,AMEX,NYSE,CBOE,PHLX,ISE,CHX,ARCA,NASDAQ,DRCTEDGE,BEX,BATS,EDGEA,BYX,IEX,EDGX,FOXRIVER,PEARL,NYSENAT,LTSE,MEMX,IBEOS,OVERNIGHT,TPLUS0,PSX\01\00\0APPLE INC\0NASDAQ\0\0Technology\0Computers\0Computers\0US/Eastern\020250324:0400-20250324:2000;20250325:0400-20250325:2000;20250326:0400-20250326:2000;20250327:0400-20250327:2000;20250328:0400-20250328:2000\020250324:0930-20250324:1600;20250325:0930-20250325:1600;20250326:0930-20250326:1600;20250327:0930-20250327:1600;20250328:0930-20250328:1600\0\0\01\0ISIN\0US0378331005\01\0\0\026,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26\0\0COMMON\00.0001\00.0001\0100\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport::connection::tests] mock -> "\0\0\0\u{b}52\01\09000\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport] <- "10\09000\0AAPL\0STK\0\00\0\0SMART\0USD\0AAPL\0NMS\0NMS\0265598\00.01\0\0ACTIVETIM,AD,ADDONT,ADJUST,ALERT,ALGO,ALLOC,AON,AVGCOST,BASKET,BENCHPX,CASHQTY,COND,CONDORDER,DARKONLY,DARKPOLL,DAY,DEACT,DEACTDIS,DEACTEOD,DIS,DUR,GAT,GTC,GTD,GTT,HID,IBKRATS,ICE,IMB,IOC,LIT,LMT,LOC,MIDPX,MIT,MKT,MOC,MTL,NGCOMB,NODARK,NONALGO,OCA,OPG,OPGREROUT,PEGBENCH,PEGMID,POSTATS,POSTONLY,PREOPGRTH,PRICECHK,REL,REL2MID,RELPCTOFS,RPI,RTH,SCALE,SCALEODD,SCALERST,SIZECHK,SMARTSTG,SNAPMID,SNAPMKT,SNAPREL,STP,STPLMT,SWEEP,TRAIL,TRAILLIT,TRAILLMT,TRAILMIT,WHATIF\0SMART,AMEX,NYSE,CBOE,PHLX,ISE,CHX,ARCA,NASDAQ,DRCTEDGE,BEX,BATS,EDGEA,BYX,IEX,EDGX,FOXRIVER,PEARL,NYSENAT,LTSE,MEMX,IBEOS,OVERNIGHT,TPLUS0,PSX\01\00\0APPLE INC\0NASDAQ\0\0Technology\0Computers\0Computers\0US/Eastern\020250324:0400-20250324:2000;20250325:0400-20250325:2000;20250326:0400-20250326:2000;20250327:0400-20250327:2000;20250328:0400-20250328:2000\020250324:0930-20250324:1600;20250325:0930-20250325:1600;20250326:0930-20250326:1600;20250327:0930-20250327:1600;20250328:0930-20250328:1600\0\0\01\0ISIN\0US0378331005\01\0\0\026,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26,26\0\0COMMON\00.0001\00.0001\0100\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport] senders: {9000: Sender { .. }}
[2025-04-01T21:53:42Z DEBUG ibapi::transport] <- "52\01\09000\0\0"
[2025-04-01T21:53:42Z DEBUG ibapi::transport] senders: {9000: Sender { .. }}
[2025-04-01T21:53:42Z ERROR ibapi::transport] error reading next message (will attempt reconnect): Error { kind: UnexpectedEof, message: "failed to fill whole buffer" }
[2025-04-01T21:53:42Z INFO  ibapi::transport] next reconnection attempt in 1s
details
[ContractDetails { contract: Contract { contract_id: 265598, symbol: "AAPL", security_type: Stock, last_trade_date_or_contract_month: "", strike: 0.0, right: "", multiplier: "", exchange: "SMART", currency: "USD", local_symbol: "AAPL", primary_exchange: "NASDAQ", trading_class: "NMS", include_expired: false, security_id_type: "", security_id: "", combo_legs_description: "", combo_legs: [], delta_neutral_contract: None, issuer_id: "", description: "" }, market_name: "NMS", min_tick: 0.01, price_magnifier: 1, order_types: ["ACTIVETIM", "AD", "ADDONT", "ADJUST", "ALERT", "ALGO", "ALLOC", "AON", "AVGCOST", "BASKET", "BENCHPX", "CASHQTY", "COND", "CONDORDER", "DARKONLY", "DARKPOLL", "DAY", "DEACT", "DEACTDIS", "DEACTEOD", "DIS", "DUR", "GAT", "GTC", "GTD", "GTT", "HID", "IBKRATS", "ICE", "IMB", "IOC", "LIT", "LMT", "LOC", "MIDPX", "MIT", "MKT", "MOC", "MTL", "NGCOMB", "NODARK", "NONALGO", "OCA", "OPG", "OPGREROUT", "PEGBENCH", "PEGMID", "POSTATS", "POSTONLY", "PREOPGRTH", "PRICECHK", "REL", "REL2MID", "RELPCTOFS", "RPI", "RTH", "SCALE", "SCALEODD", "SCALERST", "SIZECHK", "SMARTSTG", "SNAPMID", "SNAPMKT", "SNAPREL", "STP", "STPLMT", "SWEEP", "TRAIL", "TRAILLIT", "TRAILLMT", "TRAILMIT", "WHATIF"], valid_exchanges: ["SMART", "AMEX", "NYSE", "CBOE", "PHLX", "ISE", "CHX", "ARCA", "NASDAQ", "DRCTEDGE", "BEX", "BATS", "EDGEA", "BYX", "IEX", "EDGX", "FOXRIVER", "PEARL", "NYSENAT", "LTSE", "MEMX", "IBEOS", "OVERNIGHT", "TPLUS0", "PSX"], under_contract_id: 0, long_name: "APPLE INC", contract_month: "", industry: "Technology", category: "Computers", subcategory: "Computers", time_zone_id: "US/Eastern", trading_hours: ["20250324:0400-20250324:2000", "20250325:0400-20250325:2000", "20250326:0400-20250326:2000", "20250327:0400-20250327:2000", "20250328:0400-20250328:2000"], liquid_hours: ["20250324:0930-20250324:1600", "20250325:0930-20250325:1600", "20250326:0930-20250326:1600", "20250327:0930-20250327:1600", "20250328:0930-20250328:1600"], ev_rule: "", ev_multiplier: 0.0, agg_group: 1, sec_id_list: [TagValue { tag: "ISIN", value: "US0378331005" }], under_symbol: "", under_security_type: "", market_rule_ids: ["26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26", "26"], real_expiration_date: "", last_trade_time: "", stock_type: "COMMON", cusip: "", ratings: "", desc_append: "", bond_type: "", coupon_type: "", callable: false, putable: false, coupon: 0.0, convertible: false, maturity: "", issue_date: "", next_option_date: "", next_option_type: "", next_option_partial: false, notes: "", min_size: 0.0001, size_increment: 0.0001, suggested_size_increment: 100.0 }]
[2025-04-01T21:53:42Z DEBUG ibapi::transport] released request_id 9000, requests.len()=0
[2025-04-01T21:53:42Z DEBUG ibapi::client] send_message(9001, RequestMessage { fields: ["9", "8", "9001", "0", "AAPL", "STK", "", "0", "", "", "SMART", "", "USD", "", "", "0", "", ""] })
[2025-04-01T21:53:42Z DEBUG ibapi::transport] -> "\0\0\0*9\08\09001\00\0AAPL\0STK\0\00\0\0\0SMART\0\0USD\0\0\00\0\0\0"
[2025-04-01T21:53:43Z INFO  ibapi::transport] reconnected !!!
[2025-04-01T21:53:43Z DEBUG ibapi::transport] -> "API\0\0\0\0\tv100..173"
[2025-04-01T21:53:43Z DEBUG ibapi::transport::connection::tests] mock <- "v100..173"
[2025-04-01T21:53:43Z DEBUG ibapi::transport::connection::tests] mock -> "\0\0\0+173\020250323 22:21:01 Greenwich Mean Time\0\0"
[2025-04-01T21:53:43Z DEBUG ibapi::transport] <- "173\020250323 22:21:01 Greenwich Mean Time\0\0"
[2025-04-01T21:53:43Z DEBUG ibapi::transport] -> "\0\0\0\t71\02\028\0\0"
[2025-04-01T21:53:43Z DEBUG ibapi::transport::connection::tests] mock <- "71\02\028\0\0"
[2025-04-01T21:53:43Z DEBUG ibapi::transport::connection::tests] mock -> "\0\0\0\u{10}15\01\0DU1234567\0\0"
[2025-04-01T21:53:43Z DEBUG ibapi::transport::connection::tests] mock -> "\0\0\0\u{7}9\01\01\0\0"
[2025-04-01T21:53:43Z DEBUG ibapi::transport] <- "15\01\0DU1234567\0\0"
[2025-04-01T21:53:43Z DEBUG ibapi::transport] <- "9\01\01\0\0"
[2025-04-01T21:53:43Z INFO  ibapi::transport] successfully reconnected to TWS/Gateway
[2025-04-01T21:53:43Z DEBUG ibapi::transport] reset message bus
details
[]
[2025-04-01T21:53:43Z DEBUG ibapi::client] dropping basic client
[2025-04-01T21:53:43Z DEBUG ibapi::transport] shutdown requested
[2025-04-01T21:53:43Z DEBUG ibapi::transport] released request_id 9001, requests.len()=0
[2025-04-01T21:53:43Z DEBUG ibapi::transport] cleanup thread exiting
[2025-04-01T21:53:44Z DEBUG ibapi::transport] dispatcher thread exiting
[2025-04-01T21:53:44Z DEBUG ibapi::transport::connection::tests] Client closed connection to Mock Server

```
</details>

wboayue commented on src/transport/connection/tests.rs:
```diff
@@ -0,0 +1,272 @@
+// Examples:
+// Testing connect -> request -> disconnect -> reconnect -> request
+#[test]
+fn test_mock_connection() {
```
I think this will be easier to reason about if broken into 3 tests.

* **test_establish_connection** - would test just handshake and start api
* **test_send_request** - test a single request (e.g. connect and get  contract details)
* **test_reconnect** - test ability to reconnect. simplest way to do this is with an API like `Client::managed_accounts`. I haven't gone back and added this to every API that could support this but some like `Client::managed_accounts` can automatically retry after a reconnect.

wboayue commented on src/transport/connection/tests.rs:
```diff
@@ -0,0 +1,272 @@
+    let contract = Contract::stock("AAPL");
+    client.contract_details(&contract).unwrap();
+
+    // TODO: this should wait until the client has reconnected instead of sleeping
```
If we test with one of the auto reconnect methods, we won't need a sleep.

wboayue commented on src/messages.rs:
```diff
@@ -349,6 +352,50 @@ impl ToField for OutgoingMessages {
+
+    pub fn packet(&self) -> String {
+        // prefix does not get length encoded
+        "API\0".to_owned() + &encode_packet(&self.encode())
```
I also conflated these in my implementation, but taking a second look I think these are separate things.
Logically, I think it's better to think of `API\0` as the handshake  and `v100..173` as the first message.
This will simplify the test setup to just one message type.

fhill2 replied:
The handshake also requires the absense of the leading null byte. This means we can't use [RequestMethod::encode()](https://github.com/wboayue/rust-ibapi/blob/8f2ab14beee94172e81404ae8cb040afbbac2923/src/messages.rs#L370) method on the handshake.

```txt
b'API\x00\x00\x00\x00\tv100..157'
   
//  RequestMethod { fields: vec!["v176..180"] }::encode()
// raises "Message header length invalid" in TWS / Gateway
b'API\x00\x00\x00\x00\tv100..157\0'
```

So the MockServer would have to:
- have different handling functions for the Message and Handshake (we extend read_prefix()) - both have different reading, encoding handling.
Or
- create a separate struct for the handshake and have a single handling function in the MockServer (we could also simplify the handling in the Connection too)

wboayue commented on src/transport/connection/tests.rs:
```diff
@@ -0,0 +1,272 @@
+//     responses: Vec<ResponseMessage>,
+// }
+#[derive(Clone, Debug)]
+enum Message {
```
If we treat the handshake as just `API\0` can replace this with a single struct, I like  `MockExchange`, that contains the request/replies pair.

wboayue commented on src/transport/connection/tests.rs:
```diff
@@ -0,0 +1,272 @@
+
+            match message {
+                Message::Handshake { request, responses } => {
+                    self.read_prefix();
```
Instead of a special message type, read_prefix will always be the first thing that the server does.

wboayue commented:
@fhill2 Thanks for working on this. It's coming along well.

fhill2 commented:
After working on the tests and using the Mock types I've noticed areas they can be improved:

Some tests require accurate scheduling of the writes/reads. For example, ensuring waiters return the same response if a request is made before the reconnect, and during the reconnect. 

The tests also depend on external resources. Currently each test requires a socket with a localhost remote.

I've made some changes to the Connection that fixes these 2 problems:

I've moved the contents of TcpMessageBus `start_dispatcher_thread()` into a separate method so it can be called from the main thread. Executing this on the main thread in the test between writes guarantees correct scheduling.

```rust
fn test_connection_no_threads() -> Result<(), Box<dyn std::error::Error>> {
    
    ....

    bus.dispatch(server_version)?;

    let subscription = bus.send_request(9000, &packet)?;

    bus.dispatch(server_version)?;
    bus.dispatch(server_version)?;

    let result = subscription.next().unwrap();

    Ok(())
}
```

Also, the std::net::TcpStream is mocked so there are no external calls. To accomplish this the Connection and TcpMessageBus are generic over trait `Stream`, and a struct that impl `Stream` is passed to the Connection instead. I've named this `TcpSocket` (for lack of a better name), and `MockSocket` is the replacement version of this struct for testing.

`Io` and `Reconnect` are supertraits of `Stream` - However these could be combined into a single trait.

```rust
pub trait Stream: Io + Reconnect + Sync + Send + 'static + std::fmt::Debug {}
pub trait Reconnect {
    fn reconnect(&self) -> Result<(), Error>;
}
pub trait Io {
    fn read_exact(&self, buf: &mut [u8]) -> Result<(), Error>;
    fn write_all(&self, buf: &[u8]) -> Result<(), Error>;
}
```

I've also refined based on your previous comments:
- removed RequestHandshake struct, and converted the Message enum to a Message struct.  
- the mock always checks if the [message contains the prefix](https://github.com/fhill2/rust-ibapi/blob/0a17a5c2612ed046c5d35a7fcc07642c405961b6/src/transport/connection/tests.rs#L35)
- Message struct renamed to MockExchange

Let me know your thoughts on the new changes.

Currently there is a single working test using the new version [here](https://github.com/fhill2/rust-ibapi/blob/0a17a5c2612ed046c5d35a7fcc07642c405961b6/src/transport/connection/tests.rs#L162) that connects -> reconnects -> send_request.

wboayue commented on src/transport.rs:
```diff
@@ -871,56 +936,37 @@ impl Connection {
-
+        let encoded = message.encode();
+        debug!("-> {encoded:?}");
+        let length_encoded = encode_length(&encoded);
```
Nit: `encoded_with_length` might work better here. There is some ambiguity now between this being the length of the encoded message or the message encoded with length.

wboayue reviewed (approved):
This looks great, thanks @fhill2. 🚀  

I like the additional traits, they improve the modularity and testability of the MessageBus. I think we have enough to merge this PR. We can incrementally add the tests marked TODO.
